### PR TITLE
Add a check for updates method

### LINKED ISSUE #2307: Enhancement: Add check for updates and allow for autoupdating inside spyder
It would be very good to have a check for updates button under the `help` menu (i guess).

Also this check could be run on startup so the user gets a notification when an update is available and offer the install update option.

Now to handle the update we could:

1. need a restart method for spyder (#2306), 

2. If using conda or anaconda we could run the StandAlone PackageManager with a flag to update spyder which would update spyder and on finish it would relaunch it and close itself.

3. We could play with some task handling specific to each OS (seems to complicated..~!)

Comments?

### PR DESCRIPTION
**Pull request by goanpeca**
Fixes #2307 

(Note from @ccordoba12: Please leave *Fixes* instead of *Implements* because that way when I merge the PR, GH will automatically close the corresponding issue. If you don't like *Fixes*, [here](https://help.github.com/articles/closing-issues-via-commit-messages/#keywords-for-closing-issues) are the other words you can use to make this happen)

## Description
Add a method to check for updates querying the Github API, using a QThread and Worker to avoid blocking the GUI.

The method is called on startup if running from installed version and will only display a message if there is an update. If running in `DEV` mode versions it will not be triggered on startup. 

There is also an entry in the Help menu to check for updates at any moment, unlike the one on startup, this action will give feedback also if no updates are available, or if there is some internet error.

## Messages

**Ubuntu**
image

**Ubuntu**
image

**Windows**
image

## New option in the general preferences
This option is on by default and then the user can decide by checking or unchecking the checkbox.

**Windows**
image

## Checklist
- [x] Agree on location in menu
- [x] Agree on message
- [x] Add preferences option to disable this check on startup (**on** by default)
- [x] Adjust check to handle stable and dev/beta cases. 
    - If using a beta release display message if higher beta release or higher stable
    - If using a stable version check only against stable versions

### CODE REVIEW

File: spyderlib/spyder.py [hunk 1]
```diff
@@ -942,6 +948,12 @@ def create_edit_action(text, tr_text, icon_name):
             support_action = create_action(self,
                                            _("Spyder support..."),
                                            triggered=self.google_group)
+            check_updates_trig = lambda: self.check_updates(True)
+            check_updates_action = create_action(self,
```
Comment (ccordoba12) [inline spyderlib/spyder.py]: I think you should use here `self.check_updates_action` directly and remove the line below.

File: spyderlib/spyder.py [hunk 2]
```diff
@@ -2660,6 +2677,87 @@ def show_tour(self, index):
...
+        QObject.__init__(self)
+        self._parent = parent
+        self._feedback = feedback
+        self.latest_release = None
+
+    def check(self):
+        """ """
+        import json
+        try:
+            from urllib.request import urlopen
```
Comment (ccordoba12) [inline spyderlib/spyder.py]: Is this try/except needed because of differences between PY3 and PY2? If so, wouldn't it be better to use py3compat here?

File: spyderlib/spyder.py [hunk 3]
```diff
@@ -2660,6 +2677,87 @@ def show_tour(self, index):
...
+        except ImportError:
+            from urllib import urlopen
+
+        url = 'https://api.github.com/repos/spyder-ide/spyder/releases'
+
+        try:
+            page = urlopen(url)
+            data = json.loads(page.read())
+            self.latest_release = data[0]['tag_name'].replace('v', '')
+        except Exception:
```
Comment (ccordoba12) [inline spyderlib/spyder.py]: Aren't there better exceptions to catch here?

File: spyderlib/spyder.py [hunk 4]
```diff
@@ -2660,6 +2677,87 @@ def show_tour(self, index):
...
+        """
+        Check for spyder updates on github releases using a QThread.
+
+        `feedback` = False is used on startup, so only positive (updates found)
+        feedback is given. `feedback` = True is used when using the menu, and
+        gives feeback if updates are, or are not found.
+        """
+        # Disable check_updates_action
+        self.check_updates_action.setDisabled(True)
+        if self.thread_updates is not None:
```
Comment (ccordoba12) [inline spyderlib/spyder.py]: `thread_updates` is used only in this function. Why do you need to define as an attribute of the window?

File: spyderlib/spyder.py [hunk 5]
```diff
@@ -2660,6 +2677,87 @@ def show_tour(self, index):
...
+            # fail.
+            try:
+                update_available = Version(latest_release) > Version(version)
+            except Exception:
+                update_available = False
+        else:
+            update_available = False
+
+        if update_available:
+            msg = _("Spyder {0} is available.".format(latest_release))
```
Comment (ccordoba12) [inline spyderlib/spyder.py]: Please add a link here to our *Releases* page so people know where to download the update ;-)

File: spyderlib/spyder.py (hunk 1 above)
Comment (goanpeca) [inline spyderlib/spyder.py]: Ok

File: spyderlib/spyder.py (hunk 2 above)
Comment (goanpeca) [inline spyderlib/spyder.py]: Ok

File: spyderlib/spyder.py (hunk 3 above)
Comment (goanpeca) [inline spyderlib/spyder.py]: For example?

**Comment by ccordoba12** [comment]
Thanks for working on this!! It's really useful.

About your todo's:

1. I think putting the action on the Help menu is fine.
2. I also think the QMessageBox is fine, except for my little comment.

File: spyderlib/spyder.py (hunk 4 above)
Comment (goanpeca) [inline spyderlib/spyder.py]: A variable is needed to keep reference to the QThread otherwise it might segfault after leaving the method.

File: spyderlib/spyder.py (hunk 5 above)
Comment (goanpeca) [inline spyderlib/spyder.py]: The github page you mean?

This one: https://github.com/spyder-ide/spyder/releases  ?

Comment (ccordoba12) [inline spyderlib/spyder.py]: Yes, with some accompanying text saying something like: *Please go to our [Releases page](https://github.com/spyder-ide/spyder/releases) to do download Spyder <new version>*

File: spyderlib/spyder.py (hunk 4 above)
Comment (ccordoba12) [inline spyderlib/spyder.py]: ok

File: spyderlib/spyder.py (hunk 3 above)
Comment (ccordoba12) [inline spyderlib/spyder.py]: I don't know, what exceptions can occur here? I mean, is not really good practice to except on everything.

File: spyderlib/spyder.py (hunk 5 above)
Comment (goanpeca) [inline spyderlib/spyder.py]: Are we sure this is the best approach with all the different ways of installing, namely conda? pip? dmgs? the package managers?

File: spyderlib/spyder.py (hunk 3 above)
Comment (goanpeca) [inline spyderlib/spyder.py]: Ok, I will try to narrow it down

File: spyderlib/spyder.py (hunk 5 above)
Comment (ccordoba12) [inline spyderlib/spyder.py]: Good point. Then more text is needed, something like: *Please use your package manager to update Spyder or go to our ...*

**Comment by goanpeca** [comment]
@blink1073 being the closest thing to a native speaker :smile_cat:, could you check if the messages make sense to you? Thanks

**Comment by goanpeca** [comment]
@ccordoba12, what about release candidates? Will those be listed as well in the releases page? I have not taken any measures to test for those.

**Comment by blink1073** [comment]
:+1: from the 'Murican

File: spyderlib/spyder.py [hunk 6]
```diff
@@ -2660,6 +2675,133 @@ def show_tour(self, index):
...
+        title = _("Spyder updates")
+
+        if error_msg is not None:
+            msg = error_msg
+            QMessageBox.information(self, title, msg)
+        else:
+            if update_available:
+                msg = _('Spyder {0} is available. <br><br>Please use your '
+                        'package manager to update Spyder or go to our '
+                        '<a href="{1}">Releases</a> page to download Spyder '
```
Comment (ccordoba12) [inline spyderlib/spyder.py]: To not be too repetitive, please change this text to

    '<a href="{1}">Releases</a> page to download this '
    'new version'

Also, please use double quotes to build this whole string. It's better because using single quotes in written English is quite common (while not so much for double ones).

File: spyderlib/spyder.py [hunk 7]
```diff
@@ -2660,6 +2675,133 @@ def show_tour(self, index):
         self.tour.set_tour(index, frames, self)
         self.tour.start_tour()
 
+    # ---- Check for Spyder Updates
+    def _check_updates_ready(self):
+        """Called by WorkerUpdates when ready"""
+        from distutils.version import StrictVersion as Version
```
Comment (ccordoba12) [inline spyderlib/spyder.py]: Instead of using this function, please use our own `check_version` one, present in `utils/programs.py`.

**Comment by ccordoba12** [comment]
@goanpeca, GH allows us to tag betas and release candidates as pre-releases.

That's exposed through the GH API, as can be seen [here](https://developer.github.com/v3/repos/releases/#response), so you should check for it when determining the latest release.

File: spyderlib/spyder.py (hunk 6 above)
Comment (ccordoba12) [inline spyderlib/spyder.py]: You know what would be cool here too? To add a link to the *Installation* page in our documentation. It could something like

*If you don't know to update Spyder, please refer to our [Installation](http://pythonhosted.org/spyder/installation.html) instructions*.

Comment (goanpeca) [inline spyderlib/spyder.py]: Ok :+1:

**Comment by goanpeca** [comment]
@ccordoba12 what should be default behavior when checking with betas or release candidates?

If the newest is a beta (or rc), ignore it and check the next one until an official non beta release is found and compare to that?

File: spyderlib/spyder.py (hunk 7 above)
Comment (goanpeca) [inline spyderlib/spyder.py]: :+1:

**Comment by Nodd** [comment]
I think that at first we should just warn about stable versions. If we include beta or rc versions, sit should be an option, disabled by default.
It should be possible to disable the check. Imagine if the installation is managed by an administrator, a user would see the popup at each start until the admin does his job. Depending on the situation it can take a while (on some Linux distros old packages are kept for a long time, think CentOS or debian stable).
Maybe packagers should have an easy option to disable this ? Checking new versions automatically when the system already has a package manager could be redundant.

**Comment by goanpeca** [comment]
1. I would prefer NOT including beta or rc in the comparison. 
2. Ok, I can add an option to have it disabled. (but it should be ON by default, otherwise it defeats the purpose)

There are too many use cases to decide when it makes sense to have it on or off... package managers in linux distros.. pips condas.. python xy... winpython etc... so I think it is still sensible to have ON by default and only check against stable versions. The message box can include a checkbox and ask the user, if he wants to continue receiving this messages... that would make it more obvious and easy to correct.

What do you think?

**Comment by Nodd** [comment]
The "off by default" part is about beta versions. Of course the check about stable versions should be on.

Including a checkbox in the message box in addition to the preferences is OK for me.

**Comment by goanpeca** [comment]
Yes, now I understood what you meant. Will make the changes soon.

@ccordoba12 just to confirm, are these the expected tags to be found in Spyder releases:

- Developments: `'3.0.0dev'`
- Alfa:  `'3.0.0alfa1'`
- Beta: `'3.0.0beta1'`
- Release candidates:  `'3.0.0rc1'`
- Stable: `'3.0.0'`

**Comment by goanpeca** [comment]
@Nodd @ccordoba12, take a look at the description for the updated look.

Should all the other boxes contain the checkbox asking? (Error box, or Uptodate Box?)

**Comment by goanpeca** [comment]
Ok so this is the logic I coded

- If running in bootstrap (DEV) then never run on startup, and when ran with the menu, it will always display up to date... (another option would be to disable this option in DEV mode)
- If the user has a stable version installed (2.3.4 for instance) then check for updates will only check against stable versions (3.0.0, 2.3.5 etc..) and will respect the entry on the preferences menu.
- If the user has a alpha, beta, or release candidate installed (3.0.0rc1...) then check for updates will check against stable version (3.0.0) or higher non stable versions (3.0.0rc2...), and it will respect the entry on the preferences menu
- Finally, If the user has a dev version installed (for whatever reason!!) then check for updates will respect the entry on the preferences menu but will always give up to date.

**Comment by goanpeca** [comment]
I also added the extra needed logic inside the Worker class, but now I think that this should go out of the `spyder.py` file, do you agree on creating a new folder called 'workers' and put this one there there? and any other eventual worker that arises?

File: spyderlib/config.py [hunk 8]
```diff
@@ -190,7 +190,9 @@ def is_ubuntu():
               'cpu_usage/timeout': 2000,
               'use_custom_margin': True,
               'custom_margin': 0,
-              'show_internal_console_if_traceback': True
+              'show_internal_console_if_traceback': True,
+
```
Comment (ccordoba12) [inline spyderlib/config.py]: Please remove this blank line

File: spyderlib/spyder.py [hunk 9]
```diff
@@ -2660,6 +2675,218 @@ def show_tour(self, index):
...
+                        "your package manager to update Spyder or go to our "
+                        "<a href=\"{1}\">Releases</a> page to download this "
+                        "new version. <br><br>If you are not sure how to "
+                        "proceed to update Spyder please refer to our "
+                        " <a href=\"{2}\">Installation</a> instructions."
+                        "" .format(latest_release, url_r, url_i))
+                box.setText(msg)
+                box.set_check_visible(True)
+                box.exec_()
+                check_updates = box.is_checked()
```
Comment (ccordoba12) [inline spyderlib/spyder.py]: I don't think we need this checkbox here

File: spyderlib/spyder.py [hunk 10]
```diff
@@ -2660,6 +2675,218 @@ def show_tour(self, index):
...
+        self.worker_updates.sig_ready.connect(self._check_updates_ready)
+        self.worker_updates.sig_ready.connect(self.thread_updates.quit)
+        self.worker_updates.moveToThread(self.thread_updates)
+        self.thread_updates.started.connect(self.worker_updates.check)
+        self.thread_updates.start()
+
+
+class MessageCheckBox(QMessageBox):
+    """
+
```
Comment (ccordoba12) [inline spyderlib/spyder.py]: This is missing its docstring

File: spyderlib/spyder.py [hunk 11]
```diff
@@ -2660,6 +2675,218 @@ def show_tour(self, index):
...
+        self.thread_updates = QThread(self)
+        self.worker_updates = WorkerUpdates(self, feedback)
+        self.worker_updates.sig_ready.connect(self._check_updates_ready)
+        self.worker_updates.sig_ready.connect(self.thread_updates.quit)
+        self.worker_updates.moveToThread(self.thread_updates)
+        self.thread_updates.started.connect(self.worker_updates.check)
+        self.thread_updates.start()
+
+
+class MessageCheckBox(QMessageBox):
```
Comment (ccordoba12) [inline spyderlib/spyder.py]: Why do you need this class? If if it's only to add the checkbox, I think it's not really necessary...

File: spyderlib/spyder.py [hunk 12]
```diff
@@ -2660,6 +2675,218 @@ def show_tour(self, index):
...
+        return self._checkbox.setChecked(value)
+
+    def set_check_visible(self, value):
+        self._checkbox.setVisible(value)
+
+    def set_checkbox_text(self, text):
+        self._checkbox.setText(text)
+
+
+class WorkerUpdates(QObject):
```
Comment (ccordoba12) [inline spyderlib/spyder.py]: You're right, I think this should go into another file. But maybe is better to call it `updates.py`

File: spyderlib/config.py (hunk 8 above)
Comment (goanpeca) [inline spyderlib/config.py]: Ok :+1:

File: spyderlib/spyder.py (hunk 9 above)
Comment (goanpeca) [inline spyderlib/spyder.py]: Based on @Nodd suggestions, having the checkbox there would make it easy for the user to disable the message in case he/she cannot update (due to admin rights.... or any other reason) spyder. I guess s not strictly necessary but it would improve user experience.

File: spyderlib/spyder.py (hunk 11 above)
Comment (goanpeca) [inline spyderlib/spyder.py]: Well, there is no Message box that has the option to include a checkbox (check previous message).

I think this could be an useful widget to have and use in other cases, I can include this as a separate file `spyderlib\widget\messageboxes.py`

File: spyderlib/spyder.py (hunk 12 above)
Comment (goanpeca) [inline spyderlib/spyder.py]: ok, but where?

I think `spyderlib\workers\updates.py` would make sense and keep it organized. This is not really a widget, and I think more workers will be needed when handling projects and version control to avoid blocking the UI.

File: spyderlib/spyder.py (hunk 9 above)
Comment (Nodd) [inline spyderlib/spyder.py]: Actually, a "Don't check updates" pushbutton would be better than a checkbox. If it can be at the opposite side thant the OK button it's even better.

File: spyderlib/spyder.py (hunk 12 above)
Comment (Nodd) [inline spyderlib/spyder.py]: For now `spyderlib\utils\updates.py` should be fine, we can always change it later when other workers are created.
Maybe we can consider the flake/pep8 checks as workers too ?

File: spyderlib/spyder.py (hunk 9 above)
Comment (goanpeca) [inline spyderlib/spyder.py]: I really dislike that idea...

File: spyderlib/spyder.py (hunk 12 above)
Comment (goanpeca) [inline spyderlib/spyder.py]: We (I, you, anyone) might forget about this at some point, instead having the `workers`directory now will be a reminder that workers (anything added to a QThread to not block the UI) should go in that place. 

Indeed, I was considering the Flake/pep8 workers as well, but they are written in a different way I think... by subclassing QThread. There are two ways of doing it. See [this](http://doc.qt.io/qt-5/qthread.html#details). And there has been many blogs/posts etc on why use one or the other.

I really like the idea of making self contained Workers and then add them to a QThread whenever needed as opposed to sublclassing a QThread.

File: spyderlib/spyder.py (hunk 9 above)
Comment (Nodd) [inline spyderlib/spyder.py]: Why is that ? Its like a "Leave me alone with your annoying popup" button.

Comment (goanpeca) [inline spyderlib/spyder.py]: If the user clicks the Check for Updates on the menu, you would get a button saying do not check for updates.... instead of displaying the actual setting and allowing for the user to adapt it right there.

From the user experience point of view I think is much better to have the checkbox.

File: spyderlib/spyder.py (hunk 12 above)
Comment (Nodd) [inline spyderlib/spyder.py]: I thought about the forgetting thing too, but it feels weird to have a directory with only one file. That said I'm fine either way !

Comment (goanpeca) [inline spyderlib/spyder.py]: I have some workers to be used with projects and version control... that is why for me it made more sense :+1:, come on @Nodd read my mind!

File: spyderlib/spyder.py (hunk 9 above)
Comment (Nodd) [inline spyderlib/spyder.py]: Oh right, didn't think about manual checking. The checkbox is better then, I agree.

Comment (goanpeca) [inline spyderlib/spyder.py]: A button will save you from one click, but would be detrimental for the user experience... I think.

Comment (ccordoba12) [inline spyderlib/spyder.py]: Ok, let's leave the checkbox then, given that @goanpeca and @Nodd think it's better ;-)

File: spyderlib/spyder.py (hunk 11 above)
Comment (ccordoba12) [inline spyderlib/spyder.py]: I don't agree with this change. For now just put it in the same file as `WorkerUpdates`. If there are more customized QMessagebox'es in the future, we can rethink that.

File: spyderlib/spyder.py (hunk 12 above)
Comment (ccordoba12) [inline spyderlib/spyder.py]: If you have some workers in the works, then I'm fine with moving this class to `spyderlib/workers/updates.py`

Comment (goanpeca) [inline spyderlib/spyder.py]: hehe Workers in the works... playing with words @ccordoba12 ?... 

Will do once I have my hands in my linux machine.

File: spyderlib/spyder.py (hunk 10 above)
Comment (ccordoba12) [inline spyderlib/spyder.py]: Please don't forget about adding this docstring

File: spyderlib/spyder.py (hunk 12 above)
Comment (ccordoba12) [inline spyderlib/spyder.py]: Nop, didn't think about a *pun intended* ;-)

**Comment by goanpeca** [comment]
I think I covered everything I intended (and more :smile_cat:) on this PR so far, thanks for the suggestions :+1:

File: spyderlib/workers/updates.py [hunk 13]
```diff
@@ -0,0 +1,148 @@
...
+
+        return (check_version(version, latest, '<'), latest_release)
+
+    def start(self):
+        """ """
+        if PY3:
+            from urllib.request import urlopen
+            from urllib.error import URLError, HTTPError
+        else:
+            from urllib2 import urlopen, URLError, HTTPError
```
Comment (ccordoba12) [inline spyderlib/workers/updates.py]: Do these imports really need to go here? I'd prefer to put them at the top of file.

File: spyderlib/workers/updates.py [hunk 14]
```diff
@@ -0,0 +1,148 @@
...
+
+        try:
+            int(last_part)
+            return True
+        except ValueError:
+            return False
+
+    def check_update_available(self, version, releases):
+        """ """
+        from spyderlib.utils.programs import check_version
```
Comment (ccordoba12) [inline spyderlib/workers/updates.py]: The same for this import

File: spyderlib/workers/updates.py [hunk 15]
```diff
@@ -0,0 +1,148 @@
...
+            releases = [r for r in releases if self.is_stable_version(r)]
+
+        latest = releases[0]
+        latest_release = releases[0]
+
+        if version.endswith('dev'):
+            return (False, latest_release)
+
+        if self.is_stable_version(latest) and \
+                version.startswith(latest) and latest != version:
```
Comment (ccordoba12) [inline spyderlib/workers/updates.py]: This line should be indented only two spaces, i.e. like this

```python
if self.is_stable_version(latest) and \
  version.startswith(latest) and latest != version:
```

File: spyderlib/workers/updates.py [hunk 16]
```diff
@@ -0,0 +1,148 @@
...
+            # Modify the last part so that i.e. 3.0.0 is bigger than 3.0.0rc2
+            parts = latest.split('.')
+            parts = parts[:-1] + [parts[-1] + 'z']
+            latest = '.'.join(parts)
```
Comment (ccordoba12) [inline spyderlib/workers/updates.py]: I think this is an ugly hack. We should fix `check_version` to work as expected

File: spyderlib/workers/updates.py [hunk 17]
```diff
@@ -0,0 +1,148 @@
...
+    def check_update_available(self, version, releases):
+        """ """
+        from spyderlib.utils.programs import check_version
+
+        if self.is_stable_version(version):
+            # Remove non stable versions from the list
+            releases = [r for r in releases if self.is_stable_version(r)]
+
+        latest = releases[0]
+        latest_release = releases[0]
```
Comment (ccordoba12) [inline spyderlib/workers/updates.py]: Why do you need to define two different variables that hold the same thing?

File: spyderlib/workers/updates.py [hunk 18]
```diff
@@ -0,0 +1,148 @@
...
+                data = page.read()
+                if not isinstance(data, str):
+                    data = data.decode()
+                data = json.loads(data)
+                releases = [item['tag_name'].replace('v', '') for item in data]
+                version = __version__
+                result = self.check_update_available(version, releases)
+                self.update_available, self.latest_release = result
+            except Exception as error:
+                #print(error)
```
Comment (ccordoba12) [inline spyderlib/workers/updates.py]: Please remove this print

File: spyderlib/workers/updates.py [hunk 19]
```diff
@@ -0,0 +1,148 @@
...
+                error_msg = _('Unable to retrieve information.')
+        except HTTPError as error:
+            error_msg = _('Unable to retrieve information.')
+            msg = 'HTTPError = ' + str(error.code)
+            #print(msg)
```
Comment (ccordoba12) [inline spyderlib/workers/updates.py]: Also this one

File: spyderlib/workers/updates.py [hunk 20]
```diff
@@ -0,0 +1,148 @@
...
+                releases = [item['tag_name'].replace('v', '') for item in data]
+                version = __version__
+                result = self.check_update_available(version, releases)
+                self.update_available, self.latest_release = result
+            except Exception as error:
+                #print(error)
+                error_msg = _('Unable to retrieve information.')
+        except HTTPError as error:
+            error_msg = _('Unable to retrieve information.')
+            msg = 'HTTPError = ' + str(error.code)
```
Comment (ccordoba12) [inline spyderlib/workers/updates.py]: This `msg` variable and the ones below are not used. Please remove them.

File: spyderlib/workers/updates.py (hunk 13 above)
Comment (goanpeca) [inline spyderlib/workers/updates.py]: No they do not need. Will move them

File: spyderlib/workers/updates.py (hunk 14 above)
Comment (goanpeca) [inline spyderlib/workers/updates.py]: ok

File: spyderlib/workers/updates.py (hunk 15 above)
Comment (goanpeca) [inline spyderlib/workers/updates.py]: ok

File: spyderlib/workers/updates.py (hunk 16 above)
Comment (goanpeca) [inline spyderlib/workers/updates.py]: Sure, I can fix it, but the hack would go in there... so it is a matter of not liking the hack... or a matter of placing it there?

File: spyderlib/workers/updates.py (hunk 18 above)
Comment (goanpeca) [inline spyderlib/workers/updates.py]: ok

File: spyderlib/workers/updates.py (hunk 20 above)
Comment (goanpeca) [inline spyderlib/workers/updates.py]: ok

File: spyderlib/workers/updates.py (hunk 13 above)
Comment (ccordoba12) [inline spyderlib/workers/updates.py]: I'd like to use something better than `LooseVersion`, which is what `check_version` is based on.

Comment (goanpeca) [inline spyderlib/workers/updates.py]: Actually LooseVersion works pretty well, except in this cases... where it does not understand that 

3.0.0 is bigger than 3.0.0rc1 etc... so the "hack" works pretty well within the context of loose version.

http://epydoc.sourceforge.net/stdlib/distutils.version.LooseVersion-class.html

Comment (ccordoba12) [inline spyderlib/workers/updates.py]: Ok, let's do that for now :-)

File: spyderlib/workers/updates.py (hunk 16 above)
Comment (Nodd) [inline spyderlib/workers/updates.py]: I think there are functions in the standard library to check and compare versions.

Comment (blink1073) [inline spyderlib/workers/updates.py]: See https://github.com/spyder-ide/spyder/blob/master/spyderlib/utils/programs.py#L204

File: spyderlib/workers/updates.py (hunk 17 above)
Comment (goanpeca) [inline spyderlib/workers/updates.py]: For the little hack that adds a `z`  to the version number... so in the case we have `3.0.0` vs `3.0.0rc2` then the comparions is made `3.0.0z` > `3.0.0rc2`, but I still need a reference to the original number without the `z`

File: spyderlib/workers/updates.py (hunk 15 above)
Comment (goanpeca) [inline spyderlib/workers/updates.py]: When I do this, pep8 complains... if I use 8 spaces, pep8 does not complain...

File: spyderlib/workers/updates.py (hunk 13 above)
Comment (goanpeca) [inline spyderlib/workers/updates.py]: Do you want me to update our `check_version` method so that it understand these cases? or we leave that for another PR ?

**Comment by goanpeca** [comment]
@ccordoba12 made suggested changes.

File: spyderlib/workers/updates.py [hunk 21]
```diff
@@ -0,0 +1,173 @@
...
+            releases = [r for r in releases if self.is_stable_version(r)]
+
+        latest_release = releases[0]
+
+        if version.endswith('dev'):
+            return (False, latest_release)
+
+        # check_version is based on LooseVersion, so a small hack is needed so
+        # that LooseVersion understands that '3.0.0' is in fact bigger than
+        # '3.0.0rc1'
```
Comment (ccordoba12) [inline spyderlib/workers/updates.py]: One last thing: please move this hack to `check_version`. It'll be more useful there :-)

File: spyderlib/workers/updates.py [hunk 22]
```diff
@@ -0,0 +1,173 @@
...
+            error_msg = _('Unable to check for updates.')
+
+        self.error = error_msg
+        self.sig_ready.emit()
+
+
+def test_msgcheckbox():
+    from spyderlib.utils.qthelpers import qapplication
+    app = qapplication()
+    app.setStyle('Plastique')
```
Comment (ccordoba12) [inline spyderlib/workers/updates.py]: Why this style line? Just curious :-)

**Comment by ccordoba12** [comment]
Another thing: are you squashing commits? I don't like that too much because it doesn't let us see the evolution of the PR in `git log`.

File: spyderlib/workers/updates.py (hunk 21 above)
Comment (goanpeca) [inline spyderlib/workers/updates.py]: Done

Comment (goanpeca) [inline spyderlib/workers/updates.py]: BEar in mind, that also, stable version will be moved to the utils.programs

File: spyderlib/workers/updates.py (hunk 22 above)
Comment (goanpeca) [inline spyderlib/workers/updates.py]: Ehh I copied from another file, and did not bother to change that :-p

**Comment by goanpeca** [comment]
For small changes, having too many commits seems messy, but for this one you are right.

I will stop squashing, when a PR is relatively big like this one.

**Comment by goanpeca** [comment]
@ccordoba12 added latest suggestions.

File: spyderlib/spyder.py [hunk 23]
```diff
@@ -2660,6 +2674,93 @@ def show_tour(self, index):
...
+        # feedback` = False is used on startup, so only positive feedback is
+        # given. `feedback` = True is used when using the menu action, and
+        # gives feeback if updates are, or are not found.
+        feedback = False
+
+        if DEV:
+            feedback = True
+        elif self.thread_updates is not None:
+            self.thread_updates.terminate()
+            feedback = True
```
Comment (ccordoba12) [inline spyderlib/spyder.py]: I don't feel at ease with this `feedback` variable. It's passed to the worker, but it doesn't do anything with it. It seems the worker just holds it, so that `_check_updates_ready` can later read it.

But wouldn't it be better to move this block to `_check_updates_ready` directly and get rid of this passing around process?

Comment (goanpeca) [inline spyderlib/spyder.py]: Yes, you are right, this are the leftovers of the original way I did it, where the method accepted parameters. Will move it

**Comment by goanpeca** [comment]
@ccordoba12 I think this one is good to go :-)

File: spyderlib/spyder.py [hunk 24]
```diff
@@ -358,7 +358,14 @@ def __init__(self, options=None):
...
         self.tour = None
         self.tours_available = None
-        
+
+        # Check for updates Thread and Worker, refereces needed to prevent
+        # segfaulting
+        self.check_updates_action = None
+        self.thread_updates = None
+        self.worker_updates = None
+        self.flag_updates_feedback = True
```
Comment (ccordoba12) [inline spyderlib/spyder.py]: Sorry, last thing :-) I don't like too much this variable name.

Suggestions:
1. `self.updates_feedback`
2. `self.give_updates_feedback`

I think any of those would convey its meaning better ;-)

Comment (goanpeca) [inline spyderlib/spyder.py]: :+1:

**Comment by goanpeca** [comment]
Next... :-)?

**Comment by ccordoba12** [comment]
Wait for Travis ;-)

**Comment by goanpeca** [comment]
;- |

**Comment by ccordoba12** [comment]
I just tested it locally and it's working as expected :-)

This is great addition, thanks a lot for working on it!!

**Comment by goanpeca** [comment]
:+1: in the release candidates we will check for sure if it is working as it should

**Comment by stonebig** [comment]
Hi all,

I discovered recently that:
- I can't rebuild Spyder in a Windows/Python3/not-anaconda environnement.
- some users (or teachers) may not feel comfortable with unsolicited web traffic like this `check for updates`, and students randomly acting on this.

Would it be possible to set this option to "no" by default ?
